### PR TITLE
CON-3553 Adding Search for Tree tables

### PR DESCRIPTION
**Pull request by StefanStefanoff**

### CODE REVIEW

File: Subscribers/Category.php [hunk 1]
```diff
@@ -70,4 +91,99 @@ private function extendTreeNodes(array $nodes)
...
+        }
+
+        $nodes = array();
+        foreach ($parents as $parent) {
+            $nodes[] = $this->createTreeNode(
+                $parent['id'],
+                $parent['description'],
+                $parent['parent'],
+                "",
+                $this->isLeaf($parent['id']),
```
Comment (mkafadarov) [inline Subscribers/Category.php]: Please use only one sql statement with "in" cause

File: Components/CategoryExtractor.php [hunk 2]
```diff
@@ -249,14 +250,232 @@ public function getStreamsByShopId($shopId)
...
+                $sql .= ' OR ';
+            }
+
+            $sql .= ' cat.category_key = ?';
+            $params[] = $item;
+        }
+
+        $sql .= " )";
+
+        $rows = Shopware()->Db()->fetchCol($sql, $params);
```
Comment (mkafadarov) [inline Components/CategoryExtractor.php]: Please make Shopware()->Db() singleton or to be injected to the constructor

**Comment by mkafadarov** [comment]
- [x] Reviewed (@mkafadarov)

File: Components/CategoryExtractor.php [hunk 3]
```diff
@@ -249,14 +257,232 @@ public function getStreamsByShopId($shopId)
...
+        $sql = 'SELECT DISTINCT(category_key), label
+                FROM `s_plugin_connect_categories` cat
+                INNER JOIN `s_plugin_connect_product_to_categories` prod_to_cat ON cat.id = prod_to_cat.connect_category_id
+                INNER JOIN `s_plugin_connect_items` attributes ON prod_to_cat.articleID = attributes.article_id
+                WHERE attributes.shop_id = ? AND attributes.stream = ?  AND (';
+
+        $params = array($shopId, $stream);
+
+        foreach ($rootCategories as $item) {
+
```
Comment (sbofirov) [inline Components/CategoryExtractor.php]: not needed empty line

File: Components/CategoryExtractor.php [hunk 4]
```diff
@@ -249,14 +257,232 @@ public function getStreamsByShopId($shopId)
...
+            $sql .= ' cat.category_key LIKE ?';
+            $params[] = $item . '%';
+        }
+
+        $sql .= " )";
+
+        $rows = $this->db->fetchPairs($sql, $params);
+
+        return $this->convertTree($this->categoryResolver->generateTree($rows), false, true);
+
```
Comment (sbofirov) [inline Components/CategoryExtractor.php]: same here

File: Components/CategoryExtractor.php [hunk 5]
```diff
@@ -249,14 +257,232 @@ public function getStreamsByShopId($shopId)
...
+        $parents = $this->getUniqueParents($rows, $parent);
+
+        $categoryKeys = array_unique(array_merge(array_keys($rows), $parents));
+
+        $result = $this->getCategoryNames($categoryKeys);
+
+        return $this->convertTree($this->categoryResolver->generateTree($result, $parent), false, true, true);
+    }
+
+    public function getUniqueParents($rows, $parent) {
```
Comment (sbofirov) [inline Components/CategoryExtractor.php]: move {

File: Components/CategoryExtractor.php [hunk 6]
```diff
@@ -291,4 +522,32 @@ private function convertTree(array $tree, $includeChildren = true)
 
         return $categories;
     }
+
+    public function getQueryCategories($query, $excludeMapped = false, $parent = "") {
```
Comment (sbofirov) [inline Components/CategoryExtractor.php]: move {

File: Subscribers/Category.php [hunk 7]
```diff
@@ -70,4 +91,99 @@ private function extendTreeNodes(array $nodes)
         return $nodes;
     }
 
+    public function getCategoriesByQuery($query, $parentId)
+    {
+
```
Comment (sbofirov) [inline Subscribers/Category.php]: remove empty line

**Pull request merged**